### PR TITLE
[Datasets] Use sampling to estimate in-memory data size for Parquet data source

### PR DESCRIPTION
**Pull request by c21**
## Why are these changes needed?

This is continuation work from https://github.com/ray-project/ray/pull/26623. The motivation is Parquet in-memory data size estimation can be off, due to the inflation when Arrow reads Parquet encoded data into memory. Parquet only records encoded data size in file footer, which in some cases (when encoding doing very well) the encoded data size is much smaller than our dataset in-memory data size when Arrow reads it.

This PR is to propose to use sampling to do a better job for in-memory data size (i.e. it will influence number of read tasks / parallelism). The core idea is to sample a small number of rows evenly from input Parquet files, read these rows into memory and calculate in-memory size, then estimate for in-memory size of all input rows.

To not impact latency performance of dataset, launch tasks in parallel to do sampling. Also introduce a config in `DatasetContext` to allow people turn this feature off in case there's regression (by default this feature is enabled).

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/tests/test_dataset_formats.py
Comment (ericl) [inline python/ray/data/tests/test_dataset_formats.py:934]: Could we also add an integration test that the size estimate is reflected in `ds.size_bytes()` properly?

File: python/ray/data/context.py
Comment (ericl) [inline python/ray/data/context.py:69]:
```suggestion
DEFAULT_DECODING_SIZE_ESTIMATION_ENABLED = True
```

How about we call this decoding size estimation? Sampling is just a implementation technique.

File: python/ray/data/context.py
Comment (ericl) [inline python/ray/data/context.py:103]: decoding_size_estimation?

File: python/ray/data/datasource/parquet_datasource.py
Comment (ericl) [inline python/ray/data/datasource/parquet_datasource.py:50]: Should we lower this given we are sampling more rows now? Maybe like to 2?

File: python/ray/data/datasource/parquet_datasource.py
Comment (ericl) [inline python/ray/data/datasource/parquet_datasource.py:305]: Won't it always be >1 since we have MIN_NUM_SAMPLES = 4? Can simplify and always use remote tasks in this case.

File: python/ray/data/datasource/parquet_datasource.py
Comment (ericl) [inline python/ray/data/datasource/parquet_datasource.py:410]: Simplify and always take _SerializedPiece?

File: python/ray/data/datasource/parquet_datasource.py
Comment (ericl) [inline python/ray/data/datasource/parquet_datasource.py:278]: This should be guarded by the context flag right?

File: python/ray/data/datasource/parquet_datasource.py
Comment (c21) [reply python/ray/data/datasource/parquet_datasource.py:305]: @ericl - it actually can be smaller than `MIN_NUM_SAMPLES`, when total number of input. files are less than `MIN_NUM_SAMPLES` (per line 289-295). I am mostly worried about adding an extra hop (i.e. task launches another task) for CUJ when user just tries example in wiki to read single parquet file (`ray.data.read_parquet(<file-name>)`). WDYT?

Comment (ericl) [reply python/ray/data/datasource/parquet_datasource.py:305]: Running a single task should be pretty lightweight--- we already run a much larger number of tasks for other operations, so I think this is fine to simplify the code paths.

Comment (c21) [reply python/ray/data/datasource/parquet_datasource.py:305]: @ericl - cool, will update.

File: python/ray/data/tests/test_dataset_formats.py
Comment (jianoaix) [inline python/ray/data/tests/test_dataset_formats.py:943]: This looks quite accurate! Have you tried some real data files?

Comment (c21) [reply python/ray/data/tests/test_dataset_formats.py:943]: @jianoaix - I also tried the mock Uber dataset in nightly test. For `s3://shuffling-data-loader-benchmarks/data/r10_000_000_000-f1000/input_data_999.parquet.snappy`, this framework estimates each row data size to be 198 bytes. The input table has 23 int64/float columns. So the estimation size is quite close.

Comment (jianoaix) [reply python/ray/data/tests/test_dataset_formats.py:943]: Nice, note those are not real data. Let me point the user to this PR and ask them to give it a try.

File: python/ray/data/datasource/parquet_datasource.py
Comment (c21) [reply python/ray/data/datasource/parquet_datasource.py]: @ericl - sure, added.

File: python/ray/data/tests/test_dataset_formats.py
Comment (c21) [reply python/ray/data/tests/test_dataset_formats.py:934]: @ericl - sure, added.

**Review by c21 — COMMENTED** [review]
Addressed all comments, @ericl this PR is ready for review again, thanks.

File: python/ray/data/datasource/parquet_datasource.py
Comment (c21) [inline python/ray/data/datasource/parquet_datasource.py:325]: This is user-facing, so please let me know if the wording is proper enough.

File: python/ray/data/datasource/parquet_datasource.py
Comment (clarkzinzow) [inline python/ray/data/datasource/parquet_datasource.py:310]: Question: I see that this is attempting to avoid data skew inherent in the file order (e.g. small files or files with more entropy within a given column happened to be at the front of the list), but would it be better to avoid data skew bias by ensuring that we're choosing files uniformly from the file size distribution, e.g. if we sorted the files by size before doing this `linspace` draw?

I'm not sure how good of a proxy file size skew is for data entropy skew in practice... If so, it would be interesting to look at this for a future PR.

File: python/ray/data/datasource/parquet_datasource.py
Comment (clarkzinzow) [inline python/ray/data/datasource/parquet_datasource.py:313]: We expect these file sampling tasks to be I/O-bound and very lightweight, should we make the CPU request relatively lightweight as well to make sure that we're not getting blocked on Ray-level queueing? cc @ericl

File: python/ray/data/datasource/parquet_datasource.py
Comment (clarkzinzow) [inline python/ray/data/datasource/parquet_datasource.py:303]: We may want to truncate the max number of file samples based on the number of CPU cores in the cluster, in concert with decreasing the CPU request of the file sampling tasks (see comment below). cc @ericl

File: python/ray/data/datasource/parquet_datasource.py
Comment (clarkzinzow) [inline python/ray/data/datasource/parquet_datasource.py:316]: I assume that this is to avoid row group bias within Parquet files; would it be better to just randomly sample a row group within each Parquet file, or are we wanting to get ~complete coverage of all row groups across Parquet files (assuming they're ~uniform)? If the latter is the goal, that sounds good.

File: python/ray/data/tests/test_dataset_formats.py
Comment (clarkzinzow) [reply python/ray/data/tests/test_dataset_formats.py:954]: @c21 We should try this with the NYC taxi dataset, it's hosted in a public Ursa Labs bucket and is used in this example: https://docs.ray.io/en/master/data/examples/nyc_taxi_basic_processing.html#reading-and-inspecting-the-data

**Review by clarkzinzow — COMMENTED** [review]
Looking good, the estimates are looking great, just a few quick questions/nits

**Review by ericl — APPROVED** [review]
LGTM

File: python/ray/data/datasource/parquet_datasource.py
Comment (ericl) [reply python/ray/data/datasource/parquet_datasource.py:313]: I think it's OK. Let's see if this becomes a problem --- if it is, we should probably add that and also some progress bars.

File: python/ray/data/datasource/parquet_datasource.py
Comment (ericl) [reply python/ray/data/datasource/parquet_datasource.py:303]: It's truncated to 20 already, which should be much below the number of CPU cores. I'd rather the statistics not be dependent on the cluster size but only the input data (would be weird if the data size changed if you autoscaled up...)

File: python/ray/data/datasource/parquet_datasource.py
Comment (ericl) [reply python/ray/data/datasource/parquet_datasource.py:316]: One goal here is determinism, so that the size_bytes() doesn't seem to randomly change each time the data is read.

Comment (clarkzinzow) [reply python/ray/data/datasource/parquet_datasource.py:316]: Yep that makes sense, sounds good. 👍

File: python/ray/data/datasource/parquet_datasource.py
Comment (clarkzinzow) [reply python/ray/data/datasource/parquet_datasource.py:303]: Yeah that makes sense, I'm mainly thinking of the getting started flow where a user is wanting to read some small Parquet files on a single machine, but I think the file sampling overhead should be negligible in that case anyway.

**Review by clarkzinzow — APPROVED** [review]
LGTM after @ericl answered some of the above questions, nice work!

File: python/ray/data/datasource/parquet_datasource.py
Comment (c21) [reply python/ray/data/datasource/parquet_datasource.py:45]: We are estimating the multiplier/ratio on top of Parquet's uncompressed/encoded data size, not Parquet compressed (on-disk) data size. So I think the wording currently should be right.

File: python/ray/data/datasource/parquet_datasource.py
Comment (c21) [reply python/ray/data/datasource/parquet_datasource.py:310]: Yeah I am not sure how good of file size being a proxy for data skew neither. In the past, I have seen files sizes are similar, but one file's columns can be much wider than another file's. I agree we can further optimize it in the future PR if we see this pattern.

File: python/ray/data/datasource/parquet_datasource.py
Comment (c21) [reply python/ray/data/datasource/parquet_datasource.py:313]: yes agree with @ericl to add this and progress bars later, if the estimation generally takes a much longer time than we expect. But thanks for @clarkzinzow bringing it up. Beside IO cost, the row group needs to be uncompressed back, so it also costs some CPU resource. In practise, I never compare the ratio between IO vs CPU when reading just a few rows into memory though.

File: python/ray/data/datasource/parquet_datasource.py
Comment (clarkzinzow) [reply python/ray/data/datasource/parquet_datasource.py:45]: Let me make sure we've got this right; we're talking about three possible compression states for Parquet data:
1. **Compressed + Parquet-encoded, on disk:** In addition to Parquet's native column encoding (dictionary, RLE, delta), the file is also compressed with e.g. Snappy or gzip.
2. **Parquet-encoded, on disk:** Just Parquet's native column encoding (dictionary, RLE, delta), no other compression.
3. **Parquet-encoded, in memory:** Just Parquet's native column encoding (dictionary, RLE, delta), no other compression.
4. **Arrow format, in memory:** Fully decompressed, including both Parquet's column encoding and any other compression, represented in Arrow's in-memory format.

It's my understanding that:
- Arrow is giving us (2),
- (2) and (3) are equivalent in terms of size in bytes,
- we don't read the full Parquet-encoded file into memory and then decode, we instead read in chunks of row groups that progressively decoded.

For those reasons, I was thinking that referring to this multiplier/ratio as being on top of "Parquet's on-disk representation" is more accurate, but I might be misunderstanding some aspect of this.

File: python/ray/data/tests/test_dataset_formats.py
Comment (jianoaix) [inline python/ray/data/tests/test_dataset_formats.py:947]: For determinism, can we read it multiple times and assert the data_size the same?

File: python/ray/data/datasource/parquet_datasource.py
Comment (c21) [reply python/ray/data/datasource/parquet_datasource.py:316]: Yeah I think determinism is better here, some prevous comments - https://github.com/ray-project/ray/pull/26623#discussion_r924086711 .

File: python/ray/data/tests/test_dataset_formats.py
Comment (clarkzinzow) [reply python/ray/data/tests/test_dataset_formats.py:943]: @jianoaix `ds.size_bytes()` won't trigger full execution of reads, it will return the pre-read (metadata + size estimation-based) sizes.

Comment (clarkzinzow) [reply python/ray/data/tests/test_dataset_formats.py:943]: Meanwhile `ds.fully_executed().size_bytes()` will force full reads: https://github.com/ray-project/ray/blob/aadd82dcbd6bb0a8083550ef3edf39c98bf08ce0/python/ray/data/dataset.py#L3404

File: python/ray/data/datasource/parquet_datasource.py
Comment (c21) [reply python/ray/data/datasource/parquet_datasource.py:45]: `estimated-arrow-in-memory-size` = `multipler-in-this-PR` * `parquet-encoded-size`.
I agree with terminology used for (1) to (4). In practice, people always compress data when writing Parquet file, so there's really no (2): Parquet encode on-disk. When talking about Parquet in-memory size, it always means Parquet encoded data. 

I can change the phrase to `Parquet's encoded representation` if that sounds better.

File: python/ray/data/tests/test_dataset_formats.py
Comment (c21) [reply python/ray/data/tests/test_dataset_formats.py:943]: Yeah I intentionly test `size_bytes()` twice here for the exactly reason @clarkzinzow mentioned. One gives us estimated size, another give us actual size.

File: python/ray/data/datasource/parquet_datasource.py
Comment (clarkzinzow) [reply python/ray/data/datasource/parquet_datasource.py:45]: Ah I see where you were coming from with this language now, got it. I'd personally lean towards `Parquet's encoded representation` since it being on-disk or in-memory shouldn't matter, the multiplier is applied to a storage-agnostic "number of bytes for Parquet-encoded data", but feel free to leave it as is since that should be implied with "Parquet in-memory representation"!

File: python/ray/data/tests/test_dataset_formats.py
Comment (jianoaix) [reply python/ray/data/tests/test_dataset_formats.py:943]: I see, I think it'd be less confusing once we make read as a stage of the plan, and then execute() is just always executing the plan (as its name suggests).

Comment (c21) [reply python/ray/data/tests/test_dataset_formats.py:943]: @jianoaix - I agree. Right now `execute()` with `force_read: bool = False` as default behavior is very easy to catch people.

File: python/ray/data/datasource/parquet_datasource.py
Comment (jianoaix) [reply python/ray/data/datasource/parquet_datasource.py:45]: I also think "Parquet in-memory representation" is a bit inaccurate and can be read as just moving Parquet encoded data into memory bytes-by-bytes, or a new format for Parquet when it's in-memory. I think is a ratio of Arrow format / Parquet format, independent of storage medium.

Comment (c21) [reply python/ray/data/datasource/parquet_datasource.py:45]: Let me update to `Parquet's encoded representation` to avoid confusion. Thanks all.

File: python/ray/data/tests/test_dataset_formats.py
Comment (c21) [reply python/ray/data/tests/test_dataset_formats.py:954]: @clarkzinzow, @ericl and @jianoaix - here is the newly estimated size after this PR for the example in https://docs.ray.io/en/master/data/examples/nyc_taxi_basic_processing.html :

```
>>> ds.size_bytes()
2461596040
>>> ds.fully_executed().size_bytes()
2263031675
```

This is more accurate now (and good thing is it has some overestimation now), I will update the doc in a followup PR.

File: python/ray/data/datasource/parquet_datasource.py
Comment (c21) [reply python/ray/data/datasource/parquet_datasource.py]: updated.

File: python/ray/data/tests/test_dataset_formats.py
Comment (c21) [reply python/ray/data/tests/test_dataset_formats.py:947]: @jianoaix - sure, added.

**Review by c21 — COMMENTED** [review]
Addressed all comments.

File: python/ray/data/tests/test_dataset_formats.py
Comment (jianoaix) [reply python/ray/data/tests/test_dataset_formats.py:954]: Wow, this is super close!

**Review by jianoaix — APPROVED** [review]

**Pull request merged**
